Validation/HGCalValidation/plugins/HGCalHitCalibration.cc

```diff
@@ -0,0 +1,301 @@
+{
+
+  //size should be HGC layers 52 is enough
+  Energy_layer_calib.clear();
+  Energy_layer_calib_fraction.clear();
+  for(unsigned int ij=0; ij<60; ++ij){
+    Energy_layer_calib.push_back(0.);
+    Energy_layer_calib_fraction.push_back(0.);
+  }
+
+  recHitTools.getEventSetup(iSetup);
+
+  edm::Handle<HGCRecHitCollection> recHitHandleEE;
+  edm::Handle<HGCRecHitCollection> recHitHandleFH;
+  edm::Handle<HGCRecHitCollection> recHitHandleBH;
+  
+  edm::Handle<std::vector<CaloParticle> > caloParticleHandle;
+  iEvent.getByToken(_caloParticles, caloParticleHandle);
+  const std::vector<CaloParticle>& caloParticles = *caloParticleHandle;
+  
+  //make a map detid-rechit
+  std::map<DetId,const HGCRecHit*> hitmap;
+  switch(algo){
+  case 1:
+    {
+      iEvent.getByToken(_recHitsEE,recHitHandleEE);
+      iEvent.getByToken(_recHitsFH,recHitHandleFH);
+      iEvent.getByToken(_recHitsBH,recHitHandleBH);
+      const auto& rechitsEE = *recHitHandleEE;
+      const auto& rechitsFH = *recHitHandleFH;
+      const auto& rechitsBH = *recHitHandleBH;
+      for(unsigned int i = 0; i < rechitsEE.size(); ++i){
+	hitmap[rechitsEE[i].detid()] = &rechitsEE[i];
+      }
+      for(unsigned int i = 0; i < rechitsFH.size(); ++i){
+	hitmap[rechitsFH[i].detid()] = &rechitsFH[i];
+      }
+      for(unsigned int i = 0; i < rechitsBH.size(); ++i){
+	hitmap[rechitsBH[i].detid()] = &rechitsBH[i];
+      }
+      break;
+    }
+  case 2:
+    {
+      iEvent.getByToken(_recHitsEE,recHitHandleEE);
+      const HGCRecHitCollection& rechitsEE = *recHitHandleEE;
+      for(unsigned int i = 0; i < rechitsEE.size(); i++){
+	hitmap[rechitsEE[i].detid()] = &rechitsEE[i];
+      }
+      break;
+    }
+  case 3:
+    {
+      iEvent.getByToken(_recHitsFH,recHitHandleFH);
+      iEvent.getByToken(_recHitsBH,recHitHandleBH);
+      const auto& rechitsFH = *recHitHandleFH;
+      const auto& rechitsBH = *recHitHandleBH;
+      for(unsigned int i = 0; i < rechitsFH.size(); i++){
+	hitmap[rechitsFH[i].detid()] = &rechitsFH[i];
+      }
+      for(unsigned int i = 0; i < rechitsBH.size(); i++){
+	hitmap[rechitsBH[i].detid()] = &rechitsBH[i];
+      }
+      break;
+    }
+  default:
+    break;
+  }
+
+  // loop over caloParticles
+  for (std::vector<CaloParticle>::const_iterator it_caloPart = caloParticles.begin(); it_caloPart != caloParticles.end(); ++it_caloPart) {
+    const SimClusterRefVector simClusterRefVector = it_caloPart->simClusters();
+
+    Energy_layer_calib.clear();
+    Energy_layer_calib_fraction.clear();
+    for(unsigned int ij=0; ij<60; ++ij) {
+      Energy_layer_calib.push_back(0.);
+      Energy_layer_calib_fraction.push_back(0.);
+    }
+
+    int seedDet = 0;
+    float seedEnergy = 0.;
+    int simClusterCount = 0;
+    
+    for (CaloParticle::sc_iterator it_sc = simClusterRefVector.begin(); it_sc != simClusterRefVector.end(); ++it_sc) {
+      const SimCluster simCluster = (*(*it_sc));
```

Comment:
use a reference instead of a copy

better is using something like 

for ( const auto & simCluster simClusterRefVector) {
}